### ISSUE TITLE
Composer v2 file cannot be read by v1

### ISSUE THREAD

**Issue by dereuromark** [composer/composer#9337]
If you have a composer v2 locally now and push a lock file, the server which might still have v1 would fail when aliasing is used.
Is that on purpose? Or a bug?

Your requirements could not be resolved to an installable set of packages.
```
  Problem 1
    - burzum/cakephp-service-layer dev-master requires cakephp/cakephp ^4.0 -> no matching package found.
    - burzum/cakephp-service-layer dev-master requires cakephp/cakephp ^4.0 -> no matching package found.
    - Installation request for burzum/cakephp-service-layer dev-master -> satisfiable by burzum/cakephp-service-layer[dev-master].
```

The composer.json file uses:

    "cakephp/cakephp": "dev-master as 4.1.0",

If I also update the servers to v2 it works again.
The lock file indicates that this alias was dropped "wrongly" somehow: https://github.com/dereuromark/cakephp-sandbox/commit/4bd4c2be42c4578148062e48c7de566cb716b506#diff-f37acfaa6b11f575a9a6f41a75fa73a61d0f8ebc2c9b8cddc215d8aca10e44f5L7226-L7229
Or least in a non-compatibility to v1 runs.

**Comment by tidusvn05** [comment]
maybe i also got same issue,

```
Generating optimized autoload files
> Illuminate\Foundation\ComposerScripts::postAutoloadDump
> @php artisan package:discover --ansi

In Handler.php line 40:

  Undefined variable: e
```

with v1, no problem.

**Comment by Seldaek** [comment]
@dereuromark hmm yeah there are a few minor cases like this where it may not be fully compatible. The main problem here is dev-master used to (in v1) normalize as 9999999-dev internally and now it remains dev-master, so that probably explains the problem. I think it might be fixable in v1 by renormalizing the aliases there. Can you please try to checkout the 1.10 branch and see if https://github.com/composer/composer/commit/8925b4f5969412997b1621ab9e477383ea6399db fixes it for you or not?

@tidusvn05 this seems completely unrelated.. but also we don't have a Handler.php file so most likely it's a bug in artisan, please report it as a new issue on https://github.com/laravel/framework and make sure to run composer with `-v` so you can report a full stack trace.

**Comment by sprymiker** [comment]
@Seldaek, was it supposed to be compatible?

**Comment by Seldaek** [comment]
@sprymiker we aim to keep it BC to allow switching back to composer 1 if needed during the transition period, but it's not possible to be 100% compatible.

**Comment by sprymiker** [comment]
@Seldaek I believe that compatibility cannot have grades. If something is incompatible for 0.01% it is still incompatible.

So if it should be compatible the best ways is to control compatibility.

I would propose having compatibility mode defined in composer.json so composer.lock is kept compatible with the previous major version.

Is it possible? What do you think about the idea?

P.S. We have a significant impact switching to composer v2. We cannot force everybody to switch to composer v2 instantly. So now we are thinking of patching composer.lock after updates and that looks very dirty and doable.

**Comment by Seldaek** [comment]
We can not make this compatible as the normalization of dev-master/dev-default/dev-trunk versions happens differently in 1 and 2, that's just the reality of things, I *believe* my fix in https://github.com/composer/composer/commit/503c5d8e5715f2bff7b35857ce94098d3f4717f2 makes Composer 1 able to read composer 2 lock files though, if someone could check and confirm that'd be more useful than debating the specifics of what is compatibility.

As far as I know this is the only incompatibility, and it's a very minor one, only concerning aliases of dev-master to some other version. If you don't use this you should not be concerned, nor have to patch lock files.

**Comment by Seldaek** [comment]
As for compatibility mode, I'd rather not add cruft like that, either we can make it compatible and we can just make it always compatible, or we can't and there is no mode that's gonna help fix that.

**Comment by dereuromark** [comment]
Just so that we have the details here, as well.

**Comment by Seldaek** [comment]
Ok so I confirmed my fix for 1.10 works, but it's definitely better to avoid needing the fix, so it's fixed in 2.0 as well. You can try and confirm for your use case with `composer self-update --snapshot` until I get a release done.
